internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	tombpkg "gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tombpkg.Tomb) error {
+	var details checkDetails
+
+	m.state.Lock() // always acquire locks in same order (state lock, then plan lock)
+	m.planLock.Lock()
+	plan := m.plan
+	m.planLock.Unlock()
+	err := task.Get(checkDetailsAttr, &details)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for perform-check task %q: %v", task.ID(), err)
+	}
+	config, ok := plan.Checks[details.Name]
+	if !ok {
+		// Check no longer exists in plan.
+		return nil
+	}
```

Comment:
How important is this? Can we not leave the Plan changed handler to manage check removals / modifications exclusively? I have not thought this through, but since the change is always aborted if the plan changes, I think I would fire off a change with whatever it needs from the change included in the task payload.

Reply:
This is an interesting idea that I did wonder about when coding it. It'd certainly be nice conceptually, but it means that (potentially sizeable) check configuration has to be serialized in the state JSON along with the task (and I'm already concerned about the additional size/slowness of the state JSON here). I'll think about it, and see if there's some way to have non-persisted attributes on the task or something.

Reply:
Then again, I just discovered `exec` changes are even bigger! https://github.com/canonical/pebble/issues/411 -- that's even worse...

Reply:
It looks like there's `TaskRunner.AddCleanup`, which allows you to call a function by task kind to clean stuff up whenever the change completes. We could use this to delete the config data -- IMO it's not that interesting for introspection after the task completes.

Alternatively we could store this data as non-persistent: there's no way to do that on the task or change, but there is at the state level via `State.Cache` (set/delete) and `State.Cached` (get). (And then you can delete that via an `AddCleanup` hook, just to avoid it staying in memory.) Snapd uses keys like `type remodelCtxKey struct { chgID string }`. I'm definitely tempted to do this via `State.Cache` -- I'll think further about this.

Reply:
For what it's worth, I'm now doing this -- I'm using `State.Cache` to store the check config (non-persistently) with a cleanup handler to remove it from the cache. So the config data won't pollute the persisted JSON, but it still gets the simplification of avoiding `m.plan` and `m.planLock` (and them potentially being out of sync).